runtime/pkg/duckdbreplicator/backup.go

```diff
@@ -0,0 +1,335 @@
+package duckdbreplicator
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"log/slog"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"gocloud.dev/blob"
+	"gocloud.dev/blob/gcsblob"
+	"gocloud.dev/gcerrors"
+	"golang.org/x/sync/errgroup"
+)
+
+type BackupFormat string
+
+const (
+	BackupFormatUnknown BackupFormat = "unknown"
+	BackupFormatDB      BackupFormat = "db"
+	BackupFormatParquet BackupFormat = "parquet"
+)
+
+type BackupProvider struct {
+	bucket *blob.Bucket
+}
+
+func (b *BackupProvider) Close() error {
+	return b.bucket.Close()
+}
+
+type GCSBackupProviderOptions struct {
+	// UseHostCredentials specifies whether to use the host's default credentials.
+	UseHostCredentials         bool
+	ApplicationCredentialsJSON string
+	// Bucket is the GCS bucket to use for backups. Should be of the form `bucket-name`.
+	Bucket string
+	// BackupFormat specifies the format of the backup.
+	// TODO :: implement backup format. Fixed to DuckDB for now.
+	BackupFormat BackupFormat
+	// UnqiueIdentifier is used to store backups in a unique location.
+	// This must be set when multiple databases are writing to the same bucket.
+	UniqueIdentifier string
+}
+
+// NewGCSBackupProvider creates a new BackupProvider based on GCS.
+func NewGCSBackupProvider(ctx context.Context, opts *GCSBackupProviderOptions) (*BackupProvider, error) {
+	client, err := newClient(ctx, opts.ApplicationCredentialsJSON, opts.UseHostCredentials)
+	if err != nil {
+		return nil, err
+	}
+
+	bucket, err := gcsblob.OpenBucket(ctx, client, opts.Bucket, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open bucket %q, %w", opts.Bucket, err)
+	}
+
+	if opts.UniqueIdentifier != "" {
+		if !strings.HasSuffix(opts.UniqueIdentifier, "/") {
+			opts.UniqueIdentifier += "/"
+		}
+		bucket = blob.PrefixedBucket(bucket, opts.UniqueIdentifier)
+	}
+	return &BackupProvider{
+		bucket: bucket,
+	}, nil
+}
+
+// syncWrite syncs the write path with the backup location.
+func (d *db) syncWrite(ctx context.Context) error {
```

Comment:
Instead of `remote` how do you feel about using `backup` i.e. `pullFromBackup` or `pushToBackUp`?
Also `push` and `pull` gives me an impression that diffs will be merged. How about `syncWriteWithBackup` and `syncBackupWithWrite` ?

Reply:
I'm fine with both `remote` and `backup` but maybe slightly prefer "remote". The reason is that the bucket contents are the source of truth and the local files should be considered a copy/cache, but "backup" makes it sound like local is the source of truth. This distinction matters from the perspective of a future where we have concurrent readers on different machines.